File: python/fcn_bootstrapBySID.py
# fcn_bootstrapBySID.py

from __future__ import annotations
import numpy as np
import pandas as pd

def fcn_bootstrapBySID(T0: pd.DataFrame, N: int | None = None) -> pd.DataFrame:
    """
    Bootstrap-by-subject-ID (SID), mirroring the MATLAB function.

    Parameters
    ----------
    T0 : pd.DataFrame
        Original data with a 'sid' column (subject ID).
    N : int, optional
        Number of subjects to sample (with replacement). If None, uses the
        number of unique SIDs in T0.

    Returns
    -------
    pd.DataFrame
        Bootstrap sample with SIDs reassigned to 1..N.
    """
    # Unique subject IDs (sorted for stable reporting)
    unique_sids = np.sort(T0["sid"].unique())
    n_subjects = unique_sids.size

    # Default N
    if N is None:
        N = n_subjects

    # Sample original SIDs with replacement (uses current np.random state)
    sampled_sids = np.random.choice(unique_sids, size=N, replace=True)

    # Build bootstrap table; reassign new SIDs 1..N
    frames = []
    for i in range(N):
        subject_rows = T0[T0["sid"] == sampled_sids[i]].copy()
        subject_rows.loc[:, "sid"] = i + 1  # new sequential SID
        frames.append(subject_rows)

    T0_boot = pd.concat(frames, ignore_index=True)

    # ---- Console summary (mirrors MATLAB prints) ----
    print("Bootstrap sample created:")
    print(f"  Original subjects: {n_subjects}")
    print(f"  Sampled subjects:  {N}")
    print(f"  Total rows in original:  {len(T0)}")
    print(f"  Total rows in bootstrap: {len(T0_boot)}")
    print(f"  New subject IDs: 1 to {N}")

    # Sampling frequencies per original SID
    counts = pd.Series(sampled_sids).value_counts()
    not_sampled = n_subjects - counts.size
    sampled_once = int((counts == 1).sum())
    sampled_2plus = int((counts >= 2).sum())

    print("\nSubject sampling frequencies:")
    print(f"  Not sampled:    {not_sampled} subjects")
    print(f"  Sampled once:   {sampled_once} subjects")
    print(f"  Sampled 2+ times: {sampled_2plus} subjects")

    # ID mapping preview
    if N <= 10:
        k = N
        print(f"\nID mapping (first {k} subjects):")
        print("  New ID -> Original ID")
        for i in range(k):
            print(f"  {i+1:6d} -> {int(sampled_sids[i])}")
    else:
        print("\nID mapping (first 5 subjects):")
        print("  New ID -> Original ID")
        for i in range(5):
            print(f"  {i+1:6d} -> {int(sampled_sids[i])}")
        print(f"  ... (showing first 5 of {N})")

    return T0_boot

File: matlab/fcn_bootstrapBySID.m
function T0_boot = fcn_bootstrapBySID(T0, N)
% BOOTSTRAPBYSID - Create bootstrap sample by resampling subject IDs
%
% Inputs:
%   T0 - Original data table with columns including 'sid' (subject ID)
%   N  - Number of subjects to sample (optional, defaults to unique subjects in T0)
%
% Output:
%   T0_boot - Bootstrap sample table with N subjects (sampled with replacement)
%             Subject IDs are reassigned as sequential integers 1, 2, ..., N

% Get unique subject IDs
unique_sids = unique(T0.sid);
n_subjects = length(unique_sids);

% If N not specified, use same number of subjects as original
if nargin < 2
    N = n_subjects;
end

% Sample N subject IDs with replacement
sampled_sids = unique_sids(randi(n_subjects, N, 1));

% Initialize empty table
T0_boot = table();

% For each sampled subject ID, add all their rows to the bootstrap table
for i = 1:N
    % Get all rows for this subject
    subject_rows = T0(T0.sid == sampled_sids(i), :);
    
    % Assign new sequential subject ID
    subject_rows.sid(:) = i;
    
    % Append to bootstrap table
    T0_boot = [T0_boot; subject_rows];
end

% Display summary
fprintf('Bootstrap sample created:\n');
fprintf('  Original subjects: %d\n', n_subjects);
fprintf('  Sampled subjects: %d\n', N);
fprintf('  Total rows in original: %d\n', height(T0));
fprintf('  Total rows in bootstrap: %d\n', height(T0_boot));
fprintf('  New subject IDs: 1 to %d\n', N);

% Show frequency of each subject in bootstrap sample
sid_counts = histcounts(sampled_sids, [unique_sids; max(unique_sids)+1]);
fprintf('\nSubject sampling frequencies:\n');
fprintf('  Not sampled: %d subjects\n', sum(sid_counts == 0));
fprintf('  Sampled once: %d subjects\n', sum(sid_counts == 1));
fprintf('  Sampled 2+ times: %d subjects\n', sum(sid_counts >= 2));

% Optional: Show mapping of old to new IDs for first few subjects
if N <= 10
    fprintf('\nID mapping (first %d subjects):\n', N);
    fprintf('  New ID -> Original ID\n');
    for i = 1:N
        fprintf('  %6d -> %d\n', i, sampled_sids(i));
    end
elseif N > 10
    fprintf('\nID mapping (first 5 subjects):\n');
    fprintf('  New ID -> Original ID\n');
    for i = 1:5
        fprintf('  %6d -> %d\n', i, sampled_sids(i));
    end
    fprintf('  ... (showing first 5 of %d)\n', N);
end

end